.github/workflows/release.yml

```diff
@@ -1,20 +1,120 @@
-name: Publish GitHub Release
-
-permissions:
-  contents: write # for creating a release and uploading release artifacts
+name: release
 
 on:
-  push:
-    tags:
-      - 'v[0-9]+.[0-9]+.[0-9]+*'
+  workflow_dispatch:
+    inputs:
+      versionNumber:
+        description: 'Release version number (v#.#.#)'
+        type: string
+        required: true
+
+env:
+  CI_COMMIT_AUTHOR: hc-github-team-tf-provider-devex
+  CI_COMMIT_EMAIL: github-team-tf-provider-devex@hashicorp.com
+
+permissions:
+  # Allow creating GitHub release
+  contents: write
+  # Allow closing associated milestone
+  issues: write
 
 jobs:
+  major-version:
+    runs-on: ubuntu-latest
+    outputs:
+      version: ${{ steps.major-version.outputs.version }}
+    steps:
+      - id: major-version
+        run: echo "version=$(echo "${{ inputs.versionNumber }}" | cut -d. -f1)" >> "$GITHUB_OUTPUT"
+  changelog-version:
+    runs-on: ubuntu-latest
+    outputs:
+      version: ${{ steps.changelog-version.outputs.version }}
+    steps:
+      - id: changelog-version
+        run: echo "version=$(echo "${{ inputs.versionNumber }}" | cut -c 2-)" >> "$GITHUB_OUTPUT"
+  changelog:
+    needs: changelog-version
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@ac593985615ec2ede58e132d2e21d2b1cbd6127c # v3.3.0
+        with:
+          fetch-depth: 0
+      - name: Batch changes
+        uses: miniscruff/changie-action@7d8e4c245da691d57ca34019a0504321fa7b7e07 # v1.0.0
+        with:
+          version: latest
+          args: batch ${{ needs.changelog-version.outputs.version }}
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+      - name: Merge changes
+        uses: miniscruff/changie-action@7d8e4c245da691d57ca34019a0504321fa7b7e07 # v1.0.0
+        with:
+          version: latest
+          args: merge
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+      - name: Git push changelog
+        run: |
+          git config --global user.name "${{ env.CI_COMMIT_AUTHOR }}"
+          git config --global user.email "${{ env.CI_COMMIT_EMAIL }}"
+          git add .
+          git commit -a -m "Update changelog"
+          git push
+  update-package-version:
+    needs: changelog
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@ac593985615ec2ede58e132d2e21d2b1cbd6127c # v3.3.0
+        with:
+          fetch-depth: 0
+      - name: Set up Node.js
+        uses: actions/setup-node@v3
+        with:
+          node-version: 18
+      - name: Update package version
+        run: npm version "${{ inputs.versionNumber }}" --git-tag-version false
+      - name: Git push
+        run: |
+          git config --global user.name "${{ env.CI_COMMIT_AUTHOR }}"
+          git config --global user.email "${{ env.CI_COMMIT_EMAIL }}"
+          git pull
```

Comment:
Nit: I think we can skip a pull as there should be no remote changes

```suggestion
```

Reply:
If I remember correctly, since all jobs are independent of each other and the checkout action defaults to checking out the latest commit at the time that the workflow was triggered (not the job), there should be remote changes from the previous jobs in this workflow.

Reply:
@SBGoods Yeah that's correct looking at `actions/checkout`'s documentation: https://github.com/actions/checkout#usage

> ```yaml
>     # The branch, tag or SHA to checkout. When checking out the repository that
>     # triggered a workflow, this defaults to the reference or SHA for that event.
>     # Otherwise, uses the default branch.
>     ref: ''
> ```

I was having some issues with `git pull` using our service token in this PR: https://github.com/hashicorp/terraform-provider-cloudinit/pull/97/files, and I think this explains why.

I initially removed the `git pull` but that'll tag on the wrong commit now. Will look for a better answer and create a PR for cloudinit

Reply:
Maybe it'd be clearest to set the `ref` input explicitly with a comment explaining that even with a `workflow_dispatch` workflow trigger, GitHub Actions will persist the original default branch commit as the Git reference (if that's what is happening)?

Reply:
As mentioned in that issue above, if we want to use the built-in `actions/checkout` property `ref`, we can pass it `${{ github.ref }}` and you'll get the latest possible from that branch no matter what SHA triggered the workflow:

Test action run: https://github.com/hashicorp/terraform-devex-repos/actions/runs/4235926468/jobs/7360142253
Workflow definition: https://github.com/hashicorp/terraform-devex-repos/blob/48858f18312864e4da4f06ef753547447101b77a/.github/workflows/test.yml#L49